controllers/teleport_master_controller/teleport_master_controller.cpp

```diff
@@ -0,0 +1,144 @@
+/*
+ * This file is part of the NUbots Codebase.
+ *
+ * The NUbots Codebase is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The NUbots Codebase is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with the NUbots Codebase.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright 2021 NUbots <nubots@nubots.net>
+ */
+ 
+// Special thanks to Hamburg Bit-Bots for ideas for this controller
+// https://github.com/bit-bots/wolfgang_robot/blob/feature/recognition/wolfgang_webots_sim/src/wolfgang_webots_sim/webots_camera_controller.py
+
+#include <array>
+#include <fstream>
+#include <iostream>
+#include <math.h>
+#include <random>
+#include <webots/Supervisor.hpp>
+
+#include "yaml-cpp/yaml.h"
+
+// AxisAngle rotations will be read from a config file and saved here
+std::vector<std::array<double, 4>> rotations;
+double minDistance;
+double zHeight;
+double xSize;
+double ySize;
+
+int main(int argc, char** argv) {
+
+    // Make sure we have the command line arguments we need. At a minimum we should have the def argument
+    // of the robot supervisor. Other arguments are def arguments of robots that should be controlled by
+    // this supervisor
+    if (argc < 2) {
+        std::cerr << "Usage: " << argv[0] << " <DEF>" << std::endl;
+        return EXIT_FAILURE;
+    }
+    // Load self-def argument which will be used to identify the robot using the webots getFromDef function
+    std::string def = argv[1];
+
+    // Load def arguments of other robots in the field into a vector
+    std::vector<webots::Node*> otherRobotsNodes;
+    // Only one supervisor instance can exist in a single controller. As such the same one will be used
+    // every time
+    webots::Supervisor supervisor = webots::Supervisor();
+    for (int i = 2; i < argc; i++) {
+        // Load the Node of the robot by def argument and add to vector
+        webots::Node* tempNode = supervisor.getFromDef(argv[i]);
+        otherRobotsNodes.push_back(tempNode);
```

Comment:
```suggestion
        otherRobotsNodes.emplace_back(supervisor.getFromDef(argv[i]));
```

Reply:
What's the difference between emplace and push?

Reply:
A list of elements has memory allocated to it. When you use push what it does is it makes new memory, constructs an instance of your object there, then copies that memory into the memory in the list.

Emplace instead takes the arguments to the constructor and constructs the new object directly in the memory that was already there for the list.

So basically emplace means one less copy than push

Reply:
So it's a case of making it run with less strain on memory resources? Why does push even exist in that case - is there a benefit?

Reply:
Not so much strain on memory resources, for the most part (and in most circumstances) you are probably talking about a handful of bytes, although this could be important in some circumstances. There are extra computational resources involved in the copying though which is probably more costly.

There will be benefits to using push in some circumstances. I am currently too tired to actually think of a case though. You got any ideas @TrentHouliston  ?

Reply:
There are very seldom cases where you should use push rather than emplace, mostly to do with weird cases where things don't have the right constructors/operators.

The history of this is `push_back` is the original and has been around since the dinosaurs roamed the earth. `emplace_back` was added in c++11 and is better in almost every way.

Reply:
Yeah okay neat so pretty much use emplace_back unless it doesn't work and then try push_back :rofl: